### DISCUSSION TITLE
Environment reconstruction speed

### DISCUSSION THREAD

**Discussion by lyy2109** [PKU-Alignment/safety-gymnasium#88] [Q&A]
Hi，I'm a new beginner of the safety_gymnasium,and I'm starting with this code:
import safety_gymnasium
env = safety_gymnasium.make('SafetyPointCircle0-v0', render_mode='human')
obs, info = env.reset()
# Set seeds
# obs, _ = env.reset(seed=0)
terminated, truncated = False, False
ep_ret, ep_cost = 0, 0
while True:
    assert env.observation_space.contains(obs)
    act = env.action_space.sample()
    assert env.action_space.contains(act)
    # modified for Safe RL, added cost
    obs, reward, cost, terminated, truncated, info = env.step(act)
    ep_ret += reward
    ep_cost += cost
    if terminated or truncated:
        observation, info = env.reset()

However, I found that the running speed of this code is extremely slow compared to the environment in gymnasium. Is there any solution or is this phenomenon normal?

**Comment by muchvo** [comment]
It seems that the slowdown you are experiencing might be due to the use of the `human` render mode, which can indeed decrease the running speed. I suggest trying the following modification:

```python
# If you are currently using this line:
# env = safety_gymnasium.make('SafetyPointCircle0-v0', render_mode='human')

# Replace it with this:
env = safety_gymnasium.make('SafetyPointCircle0-v0')
```

This change should help in running the environment without rendering, which typically speeds up the process. Let us know if this resolves the issue or if you need further assistance!

**Reply by lyy2109** [reply]
Thank you for your reply. When I change the rendering mode, the speed will indeed increase. But my doubt is that when I set up a mujoco robot task in gymnasium, such as Humanoid-v4, it runs much faster than that in safety_gymnasium. Is there any possible solutions?

**Reply by muchvo** [reply]
You might want to try our Safe Velocity series of environments, which are derived from the Gymnasium's MuJoCo series tasks. If the performance of these environments on your machine is consistent with the speed of Gymnasium's MuJoCo environments, it would suggest that the complexity of the scenes themselves is causing the physics engine to simulate more slowly.

**Reply by muchvo** [reply]
The scenarios in our Navigation series are more complex. To address this, we offer vector environments that can run multiple environments in parallel to speed up the training process. For instructions on how to use this feature, please refer to the documentation [here](https://gymnasium.farama.org/api/vector/).

**Comment by muchvo** [comment]
Additionally, I noticed that your title mentions 'reconstruction'. I'm wondering if you're experiencing slow reset speeds. This is due to the design mechanism of the Navigation environment's code. However, currently, most Navigation environments do not encounter 'Terminated=True', meaning they would only reset once every 1000 steps. In a 1e6 step scenario, this would amount to only about 1000 resets. At present, this impact on performance is not critical, but we do plan to gradually optimize it in future versions.

**Reply by lyy2109** [reply]
Thank you! Actually my problem is the slow reset speed and I have understand your meaning

**Reply by muchvo** [reply]
Thank you for bringing this to our attention. The issue you've encountered is indeed due to the fact that each reset operation currently reconstructs a new MuJoCo instance. We acknowledge this limitation and are planning to refactor the relevant parts in upcoming versions to address this concern. 

In the meantime, if your research requires faster reset functionality, I suggest temporarily modifying the code to maintain the same instance. During the reset, you would then only need to reset or set the relevant state information. We appreciate your understanding and patience.

**Reply by lyy2109** [reply]
Besides, can you tell me how to only reset or set the relevant state information? According to the source code, the reset part contains 4 components, the first part of which is to re-establish the MuJoCo instance.But when I remove this part, the code will report an error. Thanks for your help.
![Screenshot from 2023-11-18 14-19-29](https://github.com/PKU-Alignment/safety-gymnasium/assets/95121001/33e03d96-fc3a-4302-b3d2-a3baaa9b2466)

**Reply by muchvo** [reply]
I apologize for the delayed response as I have been occupied with exams recently. You can achieve what you're looking for by directly invoking MuJoCo's underlying interfaces. Here is the documentation on [state fields](https://mujoco.readthedocs.io/en/latest/APIreference/APItypes.html#mjtstate) and the [relevant interfaces](https://mujoco.readthedocs.io/en/latest/APIreference/APIfunctions.html?highlight=state#mj-getstate). While this approach might introduce some less-than-ideal code, it should serve as a quick solution to meet your needs in the short term. However, it does require gaining some understanding of MuJoCo's lower-level operations, which can be learned from its [comprehensive documentation](https://mujoco.readthedocs.io/en/latest/index.html).

Additionally, you may need to familiarize yourself with the logic of our implementation. The code for constructing MuJoCo instances is encapsulated [here](https://github.com/PKU-Alignment/safety-gymnasium/blob/main/safety_gymnasium/world.py). The code for generating world information is primarily located within each object's [own class](https://github.com/PKU-Alignment/safety-gymnasium/tree/main/safety_gymnasium/assets) and the [random generator](https://github.com/PKU-Alignment/safety-gymnasium/blob/main/safety_gymnasium/utils/random_generator.py). 

I hope this helps, and please feel free to reach out if you have any more questions or need further assistance.

**Reply by lyy2109** [reply]
Thank for your reply!
